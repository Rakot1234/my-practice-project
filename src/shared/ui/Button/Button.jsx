import React, { PureComponent } from 'react';
import './Button.scss';
import PropTypes from 'prop-types';
import cx from 'classnames';
import Link from '../../ui/Link/Link';
import Icon from '../../ui/Icon/Icon';

class Button extends PureComponent {
    static propTypes = {
        title: PropTypes.string.isRequired,
        className: PropTypes.string,
        href: PropTypes.string,
        icon: PropTypes.string,
        size: PropTypes.oneOf(['small', 'medium', 'large']),
        color: PropTypes.oneOf(['yellow', 'red', 'gray']),
        hoverColor: PropTypes.oneOf(['black', 'sandy']),
        fontSize: PropTypes.oneOf(['medium', 'large']),
        fontColor: PropTypes.oneOf(['white', 'black']),
        onClick: PropTypes.func
    };

    static defaultProps = {
        size: 'medium',
        color: 'yellow',
        hoverColor: 'black',
        fontSize: 'medium',
        fontColor: 'white',
        className: ''
    };

    render() {
        const {
            title,
            className,
            href,
            icon,
            size,
            color,
            hoverColor,
            onClick,
            fontSize,
            fontColor
        } = this.props;
        const wrapperClasses = [
            'button', 
            className, 
            `button__color_${color}`, 
            `button__hover-color_${hoverColor}`,
            `button__size_${size}`,
            `button__font-size_${fontSize}`,
            `button__font-color_${fontColor}`
        ];

        return (
            <div
                className={cx(...wrapperClasses)}
                onClick={onClick}
            >
                {icon && <Icon icon={icon} className={cx('button__icon')} />}
                {!href && title}
                {href &&
                    <Link href={href} title={title} className={cx('button__link')} />
                }
            </div>
        );
    }
};

export default Button;
