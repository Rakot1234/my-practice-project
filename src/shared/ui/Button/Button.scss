@import '../../../styles/variables.scss';

$block: button;

.#{$block} {
    display: flex;
    align-items: center;
    justify-content: center;
    position: relative;
    box-sizing: border-box;
    padding: 0 15px;
    cursor: pointer;
    transition: all 0.3s;
    border-radius: 50px;
    border: none;
    width: 100%;

    &:focus {
        outline: none;
    }

    &:disabled {
        cursor: default;
    }

    &__link {
        display: block;
    }

    &__wrapper {
        width: 100%;

        &_disabled {
            cursor: default;
            opacity: 0.5;
        }
    }

    &__link {
        width: 100%;
    }

    &__title {
        font-weight: inherit;

        &:hover {
            text-decoration: none;
        }
    }

    &__icon {
        margin-right: 13px;
    }

    //size mods
    &_size {
        &_small {
            min-height: 32px;
        }

        &_medium {
            min-height: 35px;
        }

        &_large {
            min-height: 44px;
        }
    }

    //color mods
    &_color {
        &_yellow {
            background-color: #{$orange};
        }

        &_red {
            background-color: #{$red};
        }

        &_gray {
            background-color: #{$mediumGray};
        }

        &_black {
            background-color: #{$black};
        }
    }

    //hover color mods
    &_hover-color {
        &_black:hover {
            background-color: #{$black};

            .#{$block}__title {
                color: #{$white};
            }
        }

        &_sandy:hover {
            background-color: #{$sandy};
        }
    }

    //button font size
    &_font-size {
        &_medium .#{$block}__title {
            font-size: 14px;
        }

        &_large .#{$block}__title {
            font-size: 16px;
            @include Ubuntu;
        }
    }

    //button font color
    &_font-color {
        &_white {
            color: #{$white};

            .#{$block}__title {
                color: #{$white};
            }
        }

        &_black {
            color: #{$black};

            .#{$block}__title {
                color: #{$black};
            }

            &:hover {
                color: #{$white};
                
                .#{$block}__title {
                    color: #{$white};
                }
            }
        }
    }

    //font-weight mods

    &_font-weight {
        &_normal {
            .#{$block}__title {
                font-weight: normal;
            }
        }

        &_500 {
            .#{$block}__title {
                font-weight: 500;
            }
        }

        &_600 {
            .#{$block}__title {
                font-weight: 600;
            }
        }

        &_bold {
            .#{$block}__title {
                font-weight: bold;
            }
        }
    }
};
