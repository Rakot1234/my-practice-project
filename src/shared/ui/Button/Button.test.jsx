import React from 'react';
import { shallow } from 'enzyme';
import Button from './Button';

describe('<Button />', () => {
    it('Renders button component', () => {
        const wrapper = shallow(<Button title="Кнопка" />);
        expect(wrapper).toMatchSnapshot();
    });

    it('Calls event on click', () => {
        const buttonClick = jest.fn();
        const wrapper = shallow(<Button title="Кнопка" onClick={buttonClick} />);
        wrapper.find('.button').simulate('click');
        expect(buttonClick).toHaveBeenCalled();
    });

    it('Renders as link', () => {
        const wrapper = shallow(<Button title="Кнопка" href="/" />);
        expect(wrapper.find('a')).toBeTruthy();
    });

    it('Renders as button', () => {
        const wrapper = shallow(<Button title="Кнопка" />);
        expect(wrapper.find('button')).toBeTruthy();
    });
});