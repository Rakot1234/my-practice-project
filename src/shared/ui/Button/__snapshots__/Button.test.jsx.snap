// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Renders button component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    className=""
    color="yellow"
    fontColor="white"
    fontSize="medium"
    fontWeight="normal"
    hoverColor="black"
    size="medium"
    title="Кнопка"
    type="button"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <button
        className="button button_color_yellow button_hover-color_black button_size_medium button_font-size_medium button_font-color_white button_font-weight_normal"
        type="button"
      >
        <span
          className="button__title"
        >
          Кнопка
        </span>
      </button>,
      "className": "button__wrapper",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          undefined,
          <span
            className="button__title"
          >
            Кнопка
          </span>,
        ],
        "className": "button button_color_yellow button_hover-color_black button_size_medium button_font-size_medium button_font-color_white button_font-weight_normal",
        "disabled": undefined,
        "href": undefined,
        "onClick": undefined,
        "type": "button",
      },
      "ref": null,
      "rendered": Array [
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Кнопка",
            "className": "button__title",
          },
          "ref": null,
          "rendered": "Кнопка",
          "type": "span",
        },
      ],
      "type": "button",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <button
          className="button button_color_yellow button_hover-color_black button_size_medium button_font-size_medium button_font-color_white button_font-weight_normal"
          type="button"
        >
          <span
            className="button__title"
          >
            Кнопка
          </span>
        </button>,
        "className": "button__wrapper",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            undefined,
            <span
              className="button__title"
            >
              Кнопка
            </span>,
          ],
          "className": "button button_color_yellow button_hover-color_black button_size_medium button_font-size_medium button_font-color_white button_font-weight_normal",
          "disabled": undefined,
          "href": undefined,
          "onClick": undefined,
          "type": "button",
        },
        "ref": null,
        "rendered": Array [
          undefined,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Кнопка",
              "className": "button__title",
            },
            "ref": null,
            "rendered": "Кнопка",
            "type": "span",
          },
        ],
        "type": "button",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
